### DISCUSSION TITLE
random number generation on platforms that have no dedicated hardware for that

### DISCUSSION THREAD

**Discussion by v923z** [micropython/micropython#10550] [General]
Hi all,

I would like to ask for some advice on an issue raised in https://github.com/v923z/micropython-ulab/issues/569. It seems to me that the `random` module in `numpy` mandates something like a seedless initialisation of a random number generator, and they claim that in this case " [fresh, unpredictable entropy](https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator) will be pulled from the OS" . Yeah, good luck with that on bare metal. 

Still, I find the idea of adding some functions of the random module to `ulab` cool (though, there is a slow workaround via the `vectorize` function), and I'd like to explore, what could generate something that is unpredictable, if there is no hardware for it. I would appreciate any comments. 

Zoltán

**Comment by dlech** [comment]
A couple of common sources of entropy on microcontollers are analog inputs and the drift between two independent clock sources.

**Comment by v923z** [comment]
But wouldn't that require that you have clock sources, or that the analogue inputs are not used for something else? I can't possibly add something like this to a purely numerical package.

**Reply by robert-hh** [reply]
> A couple of common sources of entropy on microcontollers are analog inputs and the drift between two independent clock sources

I looked for a while for something like this to be added to the SAMD21 devices. But these sources are slow. So the only feasible approach would be using it for seeding and salting a PRNG. Like true RNG, this is specific for each port & MCU series.

**Reply by robert-hh** [reply]
I'm wondering that ulab asks for unpredictable numbers. I would have expected that for numerical applications well distributed numbers without an obvious pattern are sufficient. Good PRNGs will do that. Unpredictable random numbers are mandatory for cryptographic applications.

**Reply by v923z** [reply]
You only need to seed once, so speed is not really an issue, I believe.

**Comment by v923z** [comment]
The problem is not the generator itself. Melissa O'Neill has a very nice overview of this in https://www.pcg-random.org/index.html, and in fact, `micropython` has a good implementation in https://github.com/micropython/micropython/blob/e8e8c7c3548623e586f44ff208ba3f7699b7f3c2/extmod/modurandom.c#L58-L66. 

But you can't avoid the problem of the seed. I do understand your point, but if you want to generate two independent random sequences, then you need the option of seeding the generators in a random way. This is what the `seed=None` keyword arguments provides in `numpy`.

**Comment by robert-hh** [comment]
The RNG in the urandom module is seeded from a TRNG  (or OS) for the ports UNIX, STM32, ESP32, ESP8266, RP2040, MIMXRT, SAMD51. This seeding happens when the module is imported or whenever you call urandom.seed() without an argument or with `None` as argument.

**Reply by v923z** [reply]
I'm not talking about `urandom`. I'm talking about a feature request that wants to insert the `numpy` module `random` in `ulab`. `numpy`'s module requires the `seed` keyword argument, and that is used when the generator is initialised, and not when the module is imported: https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator. Otherwise, you couldn't generate independent sequences.

Your list covers pretty much all `micropython` ports, with the exception of the `nrf`, and `javascript`. Is that correct? I think, one could just bail out on the `nrf` port, if the seed is not specified.

**Reply by robert-hh** [reply]
This seed init function is not present in the `nrf` port, although it could be easily implemented, since an appropriate lib function exists. I could do that. Similar should be available for the javascript port, since it runs on PCs. But that's not my domain. I consider Java only as coffee region.

**Reply by v923z** [reply]
It would be great, but if you have better things to do, just let me know, I can take care of the implementation.

**Reply by robert-hh** [reply]
Actually it requires just the few lines below in mpconfigport.h. A fucntion `rng_generate_random_word()` returning a 32bit random already exists.
```
#if MICROPY_HW_ENABLE_RNG
#define MICROPY_PY_URANDOM_SEED_INIT_FUNC (rng_generate_random_word())
long unsigned int rng_generate_random_word(void);
#endif
```

**Reply by v923z** [reply]
Uhm, would it be possible to decouple this from `mpconfigport.h`? I think I can't override this from the `ulab` side, so if someone wants to compile `micropython` with `ulab`, they can't just simply include the user module, and run with it

**Reply by robert-hh** [reply]
The `MICROPY_PY_URANDOM_SEED_INIT_FUNC` macro is required in `extmod/modurandom.c`, which cannot be changed. So it has to be located in either `mpconfigport.h` or in a `mpconfigboard.h` file. But you said that you do not use the MP random module. Then your code may call directly `rng_generate_random_word()` to get a seed value.

**Reply by v923z** [reply]
That's right. There is only one function that I could, perhaps, use, but that is so short that it's easier to copy the implementation than to make sure that it can be included there.

**Comment by jimmo** [comment]
@v923z  I'm not sure I completely follow the discussion, but it sounds like what you want is an independent RNG that behaves like micropython's built-in urandom (i.e. a PRNG like yasmarang), but has the ability to set a seed independently of urandom. Is that correct?

As @robert-hh points out, you can basically rely on urandom to provide you with a useful seed because all the major ports initialise urandom at startup from a hardware random source.

So for example, if we made `uint32_t yasmarang(void)` public in modurandom.c (and added a modurandom.h) which would allow you to obtain a "good" seed if the user doesn't provide one, and then you could implement your own PRNG+state in ulab (e.g. your own copy of yasmarang), then would that give you what you need?

**Comment by v923z** [comment]
Yes, that pretty much sums it up. It wasn't clear to me what I should do, when a particular platform doesn't have hardware RNG, and that's why I was curious to see what alternatives there are.

But beyond that, I am not sure that making `yasmarang` public would solve the problem. If you want to initialise two independent random number generators, you would do something like this (pseudocode):

```
rng1 = random.random.Generator(yasmarang())
rng2 = random.random.Generator(yasmarang())
```
However, in `urandom`, you keep track of the state of `yasmarang`, so in the two instructions above, you basically initialise `rng1` and `rng2` with seeds that are consecutive elements in `yasmarang`'s series. Is that not correct? 

Of course, if `random.random.Generator` is not based on the numbers in `yasmarang`, that would be OK.

**Reply by jimmo** [reply]
> However, in `urandom`, you keep track of the state of `yasmarang`, so in the two instructions above, you basically initialise `rng1` and `rng2` with seeds that are consecutive elements in `yasmarang`'s series. Is that not correct?

@v923z  That is conceptually what I meant, yes. I don't know what your use case is here, my point is that as long as urandom's yasmarang was seeded with a good source at start-up, why would this be a problem?

My understanding is that you have some sort of PRNG which can be optionally seeded by the user, but the default cause is a default seed. You just need _something_ that isn't e.g. "time since boot". And what I'm saying is that on MicroPython (and I believe CircuitPython too?) this will be the case. I don't see why consecutive values from yasmarang is a problem for that?

Maybe to clarify something, if we had a class Yasmarang that had the internal state (pad, n, d, dat) and implemented the yasmarang algorithm:

```py
urandom = Yasmarang(some_seed)
rng1 = Yasmarang(urandom.next())
rng2 = Yasmarang(urandom.next())
```

If I advance rng1 (i.e. `rng1.next()`) once, then at that point it is _not_ the case that

`rng1.next() == rng2.next()`

i.e. rng1 and rng2 are still independent sequences.

```c
#include <stdio.h>
#include <stdint.h>

struct yas_state_t {
    uint32_t yasmarang_pad;
    uint32_t yasmarang_n;
    uint32_t yasmarang_d;
    uint8_t yasmarang_dat;
};

uint32_t yasmarang(struct yas_state_t * s) {
    s->yasmarang_pad += s->yasmarang_dat + s->yasmarang_d * s->yasmarang_n;
    s->yasmarang_pad = (s->yasmarang_pad << 3) + (s->yasmarang_pad >> 29);
    s->yasmarang_n = s->yasmarang_pad | 2;
    s->yasmarang_d ^= (s->yasmarang_pad << 31) + (s->yasmarang_pad >> 1);
    s->yasmarang_dat ^= (char)s->yasmarang_pad ^ (s->yasmarang_d >> 8) ^ 1;

    return s->yasmarang_pad ^ (s->yasmarang_d << 5) ^ (s->yasmarang_pad >> 18) ^ (s->yasmarang_dat << 1);
}

void seed_yasmarang(struct yas_state_t * s, uint32_t seed) {
    s->yasmarang_pad = seed;
    s->yasmarang_n = 69;
    s->yasmarang_d = 233;
    s->yasmarang_dat = 0;
}

int main(int argc, char** argv) {
    struct yas_state_t urandom = { 0xeda4baba, 69, 233, 0 };
    struct yas_state_t rng1;
    struct yas_state_t rng2;
    seed_yasmarang(&rng1, yasmarang(&urandom));
    seed_yasmarang(&rng2, yasmarang(&urandom));

    printf("urandom: %08x\n", yasmarang(&urandom));
    printf("urandom: %08x\n", yasmarang(&urandom));
    printf("urandom: %08x\n", yasmarang(&urandom));
    printf("urandom: %08x\n", yasmarang(&urandom));

    printf("rng1:    %08x\n", yasmarang(&rng1));
    printf("rng1:    %08x\n", yasmarang(&rng1));
    printf("rng1:    %08x\n", yasmarang(&rng1));
    printf("rng1:    %08x\n", yasmarang(&rng1));

    printf("rng2:    %08x\n", yasmarang(&rng2));
    printf("rng2:    %08x\n", yasmarang(&rng2));
    printf("rng2:    %08x\n", yasmarang(&rng2));
    printf("rng2:    %08x\n", yasmarang(&rng2));

    return 0;
}
```

```
$ gcc -O2 -o yas yas.c
$ ./yas
urandom: bc0c7b9d
urandom: bdec5ade
urandom: a5f12d68
urandom: 978ca203
rng1:    57797dc5
rng1:    44217bdf
rng1:    2e51e4ee
rng1:    71b737f1
rng2:    03ef17be
rng2:    ab5a1a10
rng2:    85ba5503
rng2:    1d6f8cc8
```

**Reply by v923z** [reply]
Many thanks, @jimmo ! 

I don't think there is anything that `micropython` should expose, because I believe, I would have to implement the RNG in `ulab`, anyway. The problem is that the function call (not the seed, but what you called the `.next` method) costs time. So, if in `ulab` you say "give me 10 000 random numbers and place them in an array", then it doesn't make sense to pay the function call's overhead. Since `yasmarang` is so small, I think it is better just to copy that part of the code.

**Comment by robert-hh** [comment]
> all the major ports initialise urandom at startup from a hardware random source.

@jimmo This was not the case for the nrf port, So I made a little PR to enable the seeding.
@v923z When hardware seeding is enabled, you can always call `random.seed()` or `random.seed(None)` to re-seed the PRNG from the hardware source.

**Reply by v923z** [reply]
@robert-hh Thanks! I'll try to whip up the code over the weekend. Would you mind, if I pinged you to look at it?

**Reply by robert-hh** [reply]
No, just send me a note.
